crossterm_input/src/input/input.rs

```diff
@@ -223,38 +225,295 @@ impl<'stdout> TerminalInput<'stdout> {
                         }
                     }
-                    KeyEvent::OnEnter => {
-                        if value == b'\r' {
-                            break;
-                        }
+                    Some(Ok(b'[')) => {
+                        // This is a CSI sequence.
+                        parse_csi(iter)
                     }
-                    KeyEvent::OnAnyKeyPress => {
-                        break;
+                    Some(Ok(c)) => {
+                        let ch = parse_utf8_char(c, iter);
+                        InputEvent::Keyboard(KeyEvent::Alt(ch?))
                     }
-                },
-                _ => {}
+                    Some(Err(_)) => return Err(error),
+                    None => InputEvent::Keyboard(KeyEvent::Esc),
+            })
+        },
+        b'\n' | b'\r' => Ok(InputEvent::Keyboard(KeyEvent::Char('\n'))),
+        b'\t' => Ok(InputEvent::Keyboard(KeyEvent::Char('\t'))),
+        b'\x7F' => Ok(InputEvent::Keyboard(KeyEvent::Backspace)),
+        c @ b'\x01'...b'\x1A' => Ok(InputEvent::Keyboard(KeyEvent::Ctrl((c as u8 - 0x1 + b'a') as char))),
+        c @ b'\x1C'...b'\x1F' => Ok(InputEvent::Keyboard(KeyEvent::Ctrl((c as u8 - 0x1C + b'4') as char))),
+        b'\0' => Ok(InputEvent::Keyboard(KeyEvent::Null)),
+        c => {
+            Ok({
+                   let ch = parse_utf8_char(c, iter);
+                   InputEvent::Keyboard(KeyEvent::Char(ch?))
+               })
+        }
+    }
+}
+
+/// Parses a CSI sequence, just after reading ^[
+/// Returns Event::Unknown if an unrecognized sequence is found.
+fn parse_csi<I>(iter: &mut I) -> InputEvent
+where I: Iterator<Item = Result<u8>> {
+    match iter.next() {
+        Some(Ok(b'[')) => match iter.next() {
+            // NOTE (@imdaveho): cannot find when this occurs;
+            // having another '[' after ESC[ not a likely scenario
+            Some(Ok(val @ b'A'...b'E')) => InputEvent::Keyboard(KeyEvent::F(1 + val - b'A')),
+            _ => InputEvent::Unknown,
+        },
+        Some(Ok(b'D')) => InputEvent::Keyboard(KeyEvent::Left),
+        Some(Ok(b'C')) => InputEvent::Keyboard(KeyEvent::Right),
+        Some(Ok(b'A')) => InputEvent::Keyboard(KeyEvent::Up),
+        Some(Ok(b'B')) => InputEvent::Keyboard(KeyEvent::Down),
+        Some(Ok(b'H')) => InputEvent::Keyboard(KeyEvent::Home),
+        Some(Ok(b'F')) => InputEvent::Keyboard(KeyEvent::End),
+        Some(Ok(b'M')) => {
+            // X10 emulation mouse encoding: ESC [ CB Cx Cy (6 characters only).
+            // NOTE (@imdaveho): cannot find documentation on this
+            let mut next = || iter.next().unwrap().unwrap();
+            let cb = next() as i8 - 32;
+            // (1, 1) are the coords for upper left.
+            let cx = next().saturating_sub(32) as u16;
+            let cy = next().saturating_sub(32) as u16;
+            match cb & 0b11 {
+                0 => {
+                    if cb & 0x40 != 0 {
```

Comment:
Would be nice if we could make some `constants` for this, so we know what `0x40` is.